content/ghc-9.12-cabal-3.14/index.md

```diff
@@ -0,0 +1,126 @@
++++
+title = "GHC 9.12 & cabal 3.14 releases"
+date = 2024-12-22
+[taxonomies]
+authors = ["Hécate"]
+categories = ["Cabal", "GHC"]
+tags = ["Release"]
++++
+
+To conclude the year 2024, the GHC and Cabal teams are happy to announce the release of GHC 9.12 and cabal 3.14.
+
+<!-- more -->
+
+
+Here are some highlights:
+
+## GHC 9.12
+
+### Language Extensions
+
+This release brings its lot of new and exciting extensions:
+* [Multiline string literals][MultilineStrings] to spread a piece of text over several lines;
+* [Or-Patterns][OrPatterns], allowing you to match on several patterns in a single `case` branch:
+
+```haskell
+data Sweet = Cupcake | Liquorice | Cookie | Raisins
+
+-- Without Or-Patterns
+tasty Cupcake = True
+tasty Cookie = True
+tasty _ = False
+
+-- With Or-Patterns
+tasty (Cupcake; Cookie) = True
+tasty (Liquorice; Raisins) = False
+```
+
+* [NamedDefaults][NamedDefaults] allows you to create type defaulting declarations for literals other than for the `Num` class. This mechanism is used to make sure that in the expression `print (6 + 7)`, the multiplication gets a concrete type like `Integer`, `Int` or `Double`, instead of the vague  `Num a => a`.  
+You can now apply this mechanism to other typeclasses, like:
+
+```haskell
+default IsString (Text, String)
+```
+
+Which means that in a module with [`OverloadedStrings`][OverloadedStrings], string literals `"like this"` will default to [`Text`][Text] instead of a vague polymorphic type.
+
+### Base Library
+
+* Improvements to backtraces and exceptions. Lots of them.
+
+#### ⚠️ Deprecations
+
+* The deprecation cycle of [GHC.Pack][GHC.Pack] has come to an end. Goodbye!
+* [GHC.Desugar][GHC.Desugar] will be removed in GHC 9.14.
+
+### Code Generation
+
+* Experimental support for the [RISC-V Platform](https://gitlab.haskell.org/ghc/ghc/-/issues/16783);
+* SIMD! In the x86: Most floating-point operations and some integer vector operations are supported 128-bit vectors **without LLVM**. [Get in touch](https://matrix.to/#/#ghc:matrix.org) to help with this effort
```

Comment:
```suggestion
* SIMD! On x86, most floating-point operations and some integer vector operations are supported for 128-bit vectors **without requiring LLVM**. [Get in touch](https://matrix.to/#/#ghc:matrix.org) to help with this effort
```

Reply:
The crux of the effort was to support vector operations in the native code generator (NCG) instead of LLVM only. Whether NCG actually compiles them to SIMD instructions on x86_64 / NEON on aarch64 / etc. is somewhat secondary (also obviously very important in practice).

I'd say "Vectorised operations are now supported by GHC native code generator and no longer require a separate compilation with LLVM".

Reply:
Right, that was what I was attempting to say.

Reply:
Does the suggestion still stand? I'm not sure I understand the distinction you're making @Bodigrim, since in my understanding (and I could be very well wrong), vector operations are ultimately tied to SIMD. And the release notes item says:

> SIMD support has been added to the X86 native code generator. For the time being, only 128 bit wide vectors are supported, with most floating-point operations implemented, together with a few integer vector operations. Other operations still require the LLVM backend. Contributors welcome!